packages/delegation-items-archiver/src/delegationItemsArchiverProcessors.ts

```diff
@@ -0,0 +1,158 @@
+import { match } from "ts-pattern";
+import {
+  agreementState,
+  DelegationId,
+  DelegationV2,
+  purposeVersionState,
+} from "pagopa-interop-models";
+import { InteropHeaders } from "pagopa-interop-commons";
+import {
+  AgreementProcessClient,
+  PurposeProcessClient,
+} from "./clients/clientsProvider.js";
+import { ReadModelService } from "./readModelService.js";
+
+export const processPurposes = async ({
+  readModelService,
+  purposeProcessClient,
+  headers,
+  delegationId,
+}: {
+  readModelService: ReadModelService;
+  purposeProcessClient: PurposeProcessClient;
+  headers: InteropHeaders;
+  delegationId: DelegationId;
+}): Promise<void> => {
+  const purposes = await readModelService.getPurposes(delegationId);
+
+  await Promise.all(
+    purposes.map(async (p) => {
+      const isDeletable = p.versions.every((v) =>
+        match(v.state)
+          .with(
+            purposeVersionState.draft,
+            purposeVersionState.waitingForApproval,
+            () => true
+          )
+          .with(
+            purposeVersionState.archived,
+            purposeVersionState.rejected,
+            purposeVersionState.active,
+            purposeVersionState.suspended,
+            () => false
+          )
+          .exhaustive()
+      );
+      if (isDeletable) {
+        return purposeProcessClient.internalDeletePurposeAfterDelegationRevocation(
+          undefined,
+          {
+            params: { id: p.id },
+            queries: { delegationId },
+            headers,
+          }
+        );
+      }
+
+      const activeOrSuspendedVersion = p.versions.find((v) =>
+        match(v.state)
+          .with(
+            purposeVersionState.active,
+            purposeVersionState.suspended,
+            () => true
+          )
+          .with(
+            purposeVersionState.archived,
+            purposeVersionState.rejected,
+            purposeVersionState.draft,
+            purposeVersionState.waitingForApproval,
+            () => false
+          )
+          .exhaustive()
+      );
+      if (activeOrSuspendedVersion) {
+        return purposeProcessClient.internalArchivePurposeVersionAfterDelegationRevocation(
+          undefined,
+          {
+            params: {
+              purposeId: p.id,
+              versionId: activeOrSuspendedVersion.id,
+            },
+            queries: { delegationId },
+            headers,
+          }
+        );
+      }
+
+      return Promise.resolve();
```

Comment:
Minor: there should be no need for `Promise.resolve()` to be returned. A function marked as `async` always returns a Promise